File: templates/_composition/MVVMLight/Project/Services/NavigationServiceEx.cs
using System;
using System.Collections.Generic;
using System.Linq;
using Windows.UI.Xaml;
using Windows.UI.Xaml.Controls;
using Windows.UI.Xaml.Media.Animation;

namespace Param_RootNamespace.Services
{
    public class NavigationServiceEx
    {
        private readonly Dictionary<string, Type> _pages = new Dictionary<string, Type>();

        private Frame _frame;

        public Frame Frame
        {
            get
            {
                if (_frame == null)
                {
                    _frame = Window.Current.Content as Frame;
                }

                return _frame;
            }

            set
            {
                _frame = value;
            }
        }

        public bool CanGoBack => Frame.CanGoBack;

        public bool CanGoForward => Frame.CanGoForward;

        public void GoBack() => Frame.GoBack();

        public void GoForward() => Frame.GoForward();

        public bool Navigate(string pageKey, object parameter = null, NavigationTransitionInfo infoOverride = null)
        {
            lock (_pages)
            {
                if (!_pages.ContainsKey(pageKey))
                {
                    throw new ArgumentException($"Page not found: {pageKey}. Did you forget to call NavigationService.Configure?", "pageKey");
                }

                var navigationResult = Frame.Navigate(_pages[pageKey], parameter, infoOverride);
                return navigationResult;
            }
        }

        public void Configure(string key, Type pageType)
        {
            lock (_pages)
            {
                if (_pages.ContainsKey(key))
                {
                    throw new ArgumentException($"The key {key} is already configured in NavigationService");
                }

                if (_pages.Any(p => p.Value == pageType))
                {
                    throw new ArgumentException($"This type is already configured with key {_pages.First(p => p.Value == pageType).Key}");
                }

                _pages.Add(key, pageType);
            }
        }

        public string GetNameOfRegisteredPage(Type page)
        {
            lock (_pages)
            {
                if (_pages.ContainsValue(page))
                {
                    return _pages.FirstOrDefault(p => p.Value == page).Key;
                }
                else
                {
                    throw new ArgumentException($"The page '{page.Name}' is unknown by the NavigationService");
                }
            }
        }
    }
}

File: templates/_composition/MVVMLight/ProjectVB/Services/NavigationServiceEx.vb
Imports System
Imports System.Collections.Generic
Imports System.Linq
Imports Windows.UI.Xaml
Imports Windows.UI.Xaml.Controls
Imports Windows.UI.Xaml.Media.Animation

Namespace Services
    Public Class NavigationServiceEx

        Private ReadOnly _pages As New Dictionary(Of String, Type)()

        Private Shared _frame As Frame

        Public Property Frame() As Frame
            Get
              If _frame Is Nothing Then
                  _frame = TryCast(Window.Current.Content, Frame)
              End If

              Return _frame
            End Get
            Set
                _frame = value
            End Set
        End Property

        Public ReadOnly Property CanGoBack() As Boolean
            Get
                Return Frame.CanGoBack
            End Get
        End Property

        Public ReadOnly Property CanGoForward() As Boolean
            Get
                Return Frame.CanGoForward
            End Get
        End Property

        Public Sub GoBack()
            Frame.GoBack()
        End Sub

        Public Sub GoForward()
            Frame.GoForward()
        End Sub

        Public Function Navigate(pageKey As String, Optional parameter As Object = Nothing, Optional infoOverride As NavigationTransitionInfo = Nothing) As Boolean
            Dim page As Type
            SyncLock _pages
                If Not _pages.TryGetValue(pageKey, page) Then
                    Throw New ArgumentException("Page not found: {pageKey}. Did you forget to call NavigationService.Configure?", "pageKey")
                End If
            End SyncLock
            Dim navigationResult = Frame.Navigate(page, parameter, infoOverride)
            Return navigationResult
        End Function


        Public Sub Configure(key As String, pageType As Type)
            SyncLock _pages
                If _pages.ContainsKey(key) Then
                    Throw New ArgumentException("The key {key} is already configured in NavigationService")
                End If

                If _pages.Any(Function(p) p.Value.Equals(pageType)) Then
                    Throw New ArgumentException("This type is already configured with key {_pages.First(p => p.Value == pageType).Key}")
                End If

                _pages.Add(key, pageType)
            End SyncLock
        End Sub

        Public Function GetNameOfRegisteredPage(page As Type) As String
            SyncLock _pages
                If _pages.ContainsValue(page) Then
                    Return _pages.FirstOrDefault(Function(p) p.Value.Equals(page)).Key
                Else
                    Throw New ArgumentException("The page '{page.Name}' is unknown by the NavigationService")
                End If
            End SyncLock
        End Function
    End Class
End Namespace
